### PR TITLE
adding support for qfabric

### PR DESCRIPTION
**Pull request by dgjnpr**
the updates should allow for support of a Qfabric while not impacting support for existing supported devices.

The code returns the following facts about a Qfabric:

(Pdb) pprint(dev.facts)
{'2RE': True,
 'HOME': '/pbdata/packages',
 'RE0': {'last_reboot_reason': '0x200:normal shutdown',
         'mastership_state': 'Protocol backup',
         'model': 'QFX Routing Engine',
         'status': 'Absent',
         'up_time': '4 days, 15 hours, 45 minutes, 44 seconds'},
 'RE1': {'last_reboot_reason': '0x200:normal shutdown',
         'mastership_state': 'Protocol master',
         'model': 'QFX Routing Engine',
         'status': 'Absent',
         'up_time': '4 days, 15 hours, 46 minutes, 35 seconds'},
 'domain': None,
 'fqdn': '<removed>',
 'hostname': '<removed>',
 'ifd_style': 'SWITCH',
 'master': 'RE1',
 'model': 'Virtual Chassis',
 'personality': 'SWITCH',
 'serialnumber': '<removed>',
 'switch_style': 'VLAN',
 'vc_capable': False,
 'version': '13.2X52-D20.51',
 'version_info': junos.version_info(major=(13, 2), type=X, minor=(52, 'D', 20), build=51)}

FYI I have "<removed>" host identifying data.

### CODE REVIEW

**Comment by jnpr-community-netdev** [comment]
Autobot:  Would an admin like to run functional tests?

**Comment by shermdog** [comment]
Hi David,

Thanks for the update.  We're going to have to test this *very* heavily as it touches some core facts and these changes almost always break other devices.

-Rick

**Comment by dgjnpr** [comment]
No worries Rick, I completely agree.

I’ve used this private version of PyEZ with the latest Ansible modules. Seems to be working well with MX, SRX and Qfabric. However I’m sure I’ve broken something. ;)

—
Cheers
Dg

From: Rick Sherman
Reply-To: Juniper/py-junos-eznc
Date: Wednesday, 21 January 2015 15:46
To: Juniper/py-junos-eznc
Cc: David Gethings
Subject: Re: [py-junos-eznc] adding support for qfabric (#331)


Hi David,

Thanks for the update. We're going to have to test this very heavily as it touches some core facts and these changes almost always break other devices.

-Rick

**Comment by shermdog** [comment]
@vnitinv Please take a look at this Pull Request.  This may also help to resolve #313 

Please ensure these changes function as intended and do not break any other device types.

Thanks,
-Rick

File: lib/jnpr/junos/facts/routing_engines.py [hunk 1]
```diff
@@ -38,7 +41,7 @@ def facts_routing_engines(junos, facts):
                 re_fd[factoid.replace('-', '_')] = x_f.text
 
         if 'mastership_state' in re_fd:
-            if facts[re_name]['mastership_state'] == 'master':
+            if facts[re_name]['mastership_state'].find('master') > 0:
```
Comment (vnitinv) [inline lib/jnpr/junos/facts/routing_engines.py]: @dgjnpr Can you please explain the reason for this changes. There are most chances of having mastership_state value to be 'master' itself. In that case find will give 0 and > 0 will fail.

```python
>>> facts['RE0']['mastership_state']
'master'
>>> facts['RE0']['mastership_state'].find('master')
0
>>> facts['RE0']['mastership_state'].find('master')>0
False```

File: lib/jnpr/junos/facts/routing_engines.py [hunk 2]
```diff
@@ -8,7 +8,10 @@ def facts_routing_engines(junos, facts):
         'model',
         'up-time',
         'last-reboot-reason']
-    re_info = junos.rpc.get_route_engine_information()
+    try:
+        re_info = junos.rpc.get_route_engine_information()
+    except:
+        re_info = junos.rpc.get_route_engine_information(infrastructure="FM-0")
```
Comment (vnitinv) [inline lib/jnpr/junos/facts/routing_engines.py]: Can we have 1-2 line description (doc string) above this line of code. This is very specific to Qfabric setup only. 
So can you explain this itself.

File: lib/jnpr/junos/facts/routing_engines.py (hunk 1 above)
Comment (dgjnpr) [inline lib/jnpr/junos/facts/routing_engines.py]: Hi @vnitinv ,

It's been so long I don't recall the reasoning for this change. Feel free to ignore it. ;)

File: lib/jnpr/junos/facts/routing_engines.py (hunk 2 above)
Comment (dgjnpr) [inline lib/jnpr/junos/facts/routing_engines.py]: Hi @vnitinv,

Added a doc string to the latest pull request

**Comment by vnitinv** [comment]
@dgjnpr I got one qfabric setup where we don't have chassis-module/Midplane too to figure out serial number from get-chassis-inventory rpc. So the changes will not work on many qfabric setup.

**Comment by dgjnpr** [comment]
Hi @vnitinv ,

If that's the case can we make the chassis-module/Midplane field optional? i.e. if the system has one we include it, otherwise we don't output the field?

**Pull request closed without merging**